Hh domain

dariaban opened the pull request:
Ребят, первый вариант готов. Опыта в разработке у меня никакого нет, буду рада любым замечаниям.

dariaban committed:
interface IStorageRepository -- использовала Shared Preferences для сохранения

dariaban committed:
IVacancyRepository попыталась состыковать с нашим дата слоем

dariaban committed:
IFavVacanciesRepository попыталась состыковать с нашим дата слоем

dariaban committed:
ISharingProvider создала интент

dariaban committed:
IFavVacanciesInteractor

dariaban committed:
IVacancyInteractor

dariaban committed:
ISharingInteractor

dariaban committed:
Поправила, то что от меня хотел detekt, тест на сборку не проходит, потому что нет половины необходимых файлов из дата слоя

Commit:
Merge branch 'dev' into HH-domain

john-krasinski commented:
Я думаю, в названиях классов в impl лучше убрать первую 'I, это интерфейсы так называют, чтобы легче отличать

Commit:
Объединил с дата слоем, поправил несостыковки. Перенес репозитории в пакет 'data'

- Поменял тип аргументов в репозитории БД на VacancyDetails, потому что Entity я формировал из VacancyDetails. Соображения такие - добавляем в избранное со страницы деталей вакансии, поэтому все детали в базу сохраняем. К тому же так можно смотреть избранные вакансии без интернета.

- VacancyInteractor:
для запроса словарей Areas и Industries убрал аргументы, там только locale, само сгенерится
для searchVacancies пока оставил expression:String, который потом вставится в запрос

- VacancyRepository:
пока сделал, чтобы возвращал целиком SearchVacanciesResponse, чтобы учитывать дальше пагинацию. Но наверно  лучше возвращать сразу список вакансий, как сделала Дарья, а SearchVacanciesResponse сохранить в самом репозитори и использовать в новом методе для подгрузки следующей страницы.

Commit:
Merge remote-tracking branch 'origin/HH-domain' into HH-domain

# Conflicts:
#	app/src/main/java/ru/practicum/android/diploma/data/IStorageRepositoryImpl.kt
#	app/src/main/java/ru/practicum/android/diploma/domain/api/IFavVacanciesInteractor.kt
#	app/src/main/java/ru/practicum/android/diploma/domain/api/IStorageRepository.kt
#	app/src/main/java/ru/practicum/android/diploma/domain/impl/IFavVacanciesRepositoryImpl.kt
#	app/src/main/java/ru/practicum/android/diploma/domain/impl/IVacancyInteractorImpl.kt
#	app/src/main/java/ru/practicum/android/diploma/domain/impl/IVacancyRepositoryImpl.kt

AlexandrFedoseev-1 commented on app/src/main/java/ru/practicum/android/diploma/data/IFavVacanciesRepositoryImpl.kt:
```diff
@@ -0,0 +1,107 @@
+            salaryCurrency = salary.currency ?: "",
+            experienceId = vacancy.experience?.id ?: "",
+            experienceName = vacancy.experience?.name ?: ""
+        )
```
Для этого мне на курсе рекомендовали создать отдельный маппер и передавать его в репозиторий.
 ```
class PlaylistEntityMapper {
    fun map(playlist: Playlist): PlaylistEntity {
        return PlaylistEntity(
            playlistId =  playlist.playlistId,
            name = playlist.name,
            description = playlist.description,
            imageUri = playlist.imageUri,
            tracksId = playlist.tracksId,
            tracksCount = playlist.tracksCount
        )
   

    fun map(playlist: PlaylistEntity): Playlist {
        return Playlist(
            playlistId =  playlist.playlistId,
            name = playlist.name,
            description = playlist.description,
            imageUri = playlist.imageUri,
            tracksId = playlist.tracksId,
            tracksCount = playlist.tracksCount
        )
    }
}`
```

AlexandrFedoseev-1 commented on app/src/main/java/ru/practicum/android/diploma/domain/api/IFavVacanciesInteractor.kt:
```diff
@@ -0,0 +1,11 @@
+import ru.practicum.android.diploma.data.dto.VacancyDetails
+
+interface IFavVacanciesInteractor {
+    suspend fun getFavorite(): Flow<List<VacancyDetails>>
```
Функция, возвращающая Flow, сама по себе не обязательно должна быть объявлена как suspend. В Kotlin Flow представляет собой холодный поток данных, который начинается с момента, когда на него подписываются, и при этом его сбор (collect) происходит асинхронно. То есть, сам Flow уже является ленивым и поддерживает асинхронное выполнение. У нас вызывается только flow{ }, поэтому делать suspend эти функции не нужно

AlexandrFedoseev-1 reviewed (commented):
Думаю лучше у всех Flow функций убрать suspend, т.к. в нашем случае он не нужен.

LuxuriousGerg commented on app/src/main/java/ru/practicum/android/diploma/data/IStorageRepositoryImpl.kt:
```diff
@@ -0,0 +1,25 @@
+import ru.practicum.android.diploma.data.dto.Vacancy
+import ru.practicum.android.diploma.domain.api.IStorageRepository
+
+class IStorageRepositoryImpl(private val storageSharedPreferences: SharedPreferences) : IStorageRepository {
```
как вариант на рассмотрение можно внедрить Gson через конструктор 
class IStorageRepositoryImpl(
    private val prefs: SharedPreferences,
    private val gson: Gson
) : IStorageRepository
Это по сути даёт гибкость для тестов и кастомной сериализации при необходимости

dariaban committed:
Ребят спасибо за помощь огромное!

john-krasinski commented on app/src/main/java/ru/practicum/android/diploma/domain/api/IVacancyRepository.kt:
```diff
@@ -0,0 +1,17 @@
+interface IVacancyRepository {
+    suspend fun searchVacancies(req: SearchVacanciesRequest): Flow<Resource<SearchVacanciesResponse>>
+    suspend fun getCountries(): Flow<Resource<List<Area>>>
+    suspend fun getRegion(): Flow<Resource<List<Area>>>
```
Тут хотел спросить, может не будем разграничивать на два запроса по локации, там все равно возвращаются деревья из `Area`, где корень это страна, а след. элементы это регионы, что думаете?

NoGameNoLifeProject commented on app/build.gradle.kts:
```diff
@@ -77,4 +77,7 @@ dependencies {
     androidTestImplementation(libs.uiTests.espressoCore)
     // endregion
+
+    implementation("com.google.code.gson:gson:2.11.0")
```
Дубликат

dariaban committed:
Detekt ругается

dariaban committed:
Поправила через терминал, как Юрий советовал

dariaban committed:
Еще попытка
